File: eel-core/src/main/scala/io/eels/component/csv/CsvSourceParser.scala
package io.eels.component.csv

import java.nio.file.Paths

import com.sksamuel.scalax.net.UrlParamParser
import io.eels.{SinkParser, SourceParser}
import io.eels.component.Builder

object CsvSourceParser extends SourceParser {
  val Regex = "csv:([^?].*?)(\\?.*)?".r
  override def apply(str: String): Option[Builder[CsvSource]] = str match {
    case Regex(path, params) =>
      Some(CsvSourceBuilder(path, Option(params).map(UrlParamParser.apply).getOrElse(Map.empty)))
    case _ => None
  }
}

case class CsvSourceBuilder(path: String, params: Map[String, List[String]]) extends Builder[CsvSource] {
  require(path != null, "path name cannot be null")
  override def apply(): CsvSource = new CsvSource(Paths.get(path))
}

object CsvSinkParser extends SinkParser {
  val Regex = "csv:([^?].*?)(\\?.*)?".r
  override def apply(str: String): Option[Builder[CsvSink]] = str match {
    case Regex(path, params) =>
      Some(CsvSinkBuilder(path, Option(params).map(UrlParamParser.apply).getOrElse(Map.empty)))
    case _ => None
  }
}

case class CsvSinkBuilder(path: String, params: Map[String, List[String]]) extends Builder[CsvSink] {
  require(path != null, "path name cannot be null")
  override def apply(): CsvSink = new CsvSink(Paths.get(path))
}

File: eel-core/src/main/kotlin/io/eels/component/csv/CsvSourceParser.kt
package io.eels.component.csv

import java.nio.file.Paths

import io.eels.Builder
import io.eels.SinkParser
import io.eels.SourceParser

object CsvSourceParser : SourceParser {
  val Regex = "csv:([^?].*?)(\\?.*)?"
  override fun parse(url: String): Builder<CsvSource>? = when (url) {
  //Regex -> CsvSourceBuilder(path, params)
    else -> null
  }
}

class CsvSourceBuilder(val path: String, params: Map<String, List<String>>) : Builder<CsvSource> {
  override fun invoke(): CsvSource = CsvSource(Paths.get(path))
}

object CsvSinkParser : SinkParser {
  val Regex = "csv:([^?].*?)(\\?.*)?"
  override fun parse(url: String): Builder<CsvSink>? = when (url) {
  // Regex -> Some(CsvSinkBuilder(path, Option(params).map(UrlParamParser.apply).getOrElse(Map.empty)))
    else -> null
  }
}

class CsvSinkBuilder(val path: String, params: Map<String, List<String>>) : Builder<CsvSink> {
  override fun invoke(): CsvSink = CsvSink(Paths.get(path))
}